From 910fc70626e1edaea2954e40fabb5a1ae4a1dcc6 Mon Sep 17 00:00:00 2001
From: Duy Nguyen <duy.nguyen.xa@renesas.com>
Date: Fri, 20 Jun 2025 17:36:26 +0700
Subject: [PATCH 19/19] target/rx: Reset the CPU at qemu reset time

This ensure that the CPU gets reset every time QEMU resets.

Signed-off-by: Keith Packard <keithp@keithp.com>
Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
---
 target/rx/cpu.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/target/rx/cpu.c b/target/rx/cpu.c
index 0ba0d55ab5..5a412746db 100644
--- a/target/rx/cpu.c
+++ b/target/rx/cpu.c
@@ -27,6 +27,7 @@
 #include "hw/loader.h"
 #include "fpu/softfloat.h"
 #include "tcg/debug-assert.h"
+#include "system/reset.h"
 
 static void rx_cpu_set_pc(CPUState *cs, vaddr value)
 {
@@ -129,6 +130,13 @@ static ObjectClass *rx_cpu_class_by_name(const char *cpu_model)
     return oc;
 }
 
+static void rx_cpu_reset(void *opaque)
+{
+    RXCPU *cpu = opaque;
+
+    cpu_reset(CPU(cpu));
+}
+
 static void rx_cpu_realize(DeviceState *dev, Error **errp)
 {
     CPUState *cs = CPU(dev);
@@ -142,7 +150,7 @@ static void rx_cpu_realize(DeviceState *dev, Error **errp)
     }
 
     qemu_init_vcpu(cs);
-    cpu_reset(cs);
+    qemu_register_reset(rx_cpu_reset, RX_CPU(cs));
 
     rcc->parent_realize(dev, errp);
 }
-- 
2.43.0

